### ISSUE TITLE
`createMutable` intensive testing

### ISSUE THREAD

**Issue by titoBouzout** [solidjs/solid#2112]
### Describe the bug

I've written a mutable primitive for my lib pota, collected tests for it from solid, oby, vue, + additions/variations I made up. I then ran the tests against the implementations on our libs: solid, oby and pota, to highlight the differences and possible bugs.

There are 240~ tests, I will list below the failing test for solid, around 35~. "by design" implementation details that fail won't count. 

1. object returned by function call in mutable is not observed.  https://playground.solidjs.com/anonymous/4d40b7dd-48a1-4f73-a6ca-bd98c2b8331f - ref 10 mutation: returned object by function call is mutable 

2. writing to an object with setter/getter overwrites getter.  https://playground.solidjs.com/anonymous/4c42006f-c124-4054-a74d-70974e0acbb4 - ref 12 getters: returning object

3. In solid frozen objects are made mutable, when a getter returns a frozen object it throws https://playground.solidjs.com/anonymous/4c62129a-5560-475b-b8fa-45e45a96f2b8 - 14 - getters: returning frozen object

4. setting to undefined shouldn't delete the property https://playground.solidjs.com/anonymous/6f1021b5-8e71-49a9-b376-1b89b64b010d - 22 - setting to undefined shouldnt delete the property

5. deleting key that doesn't exists triggers reactivity https://playground.solidjs.com/anonymous/2d07b152-75de-4ea9-a085-c1c5338a2c8f - 24 - delete non existent key doesnt trigger reactivity - object.keys

6.  delete key with undefined value doesnt trigger reactivity with check "in" https://playground.solidjs.com/anonymous/5c6627a9-4665-446e-96b4-31b697a77018 delete key with undefined value does trigger reactivity - in

7. attempting to set a value when its only a getter access getter https://playground.solidjs.com/anonymous/44d261b9-70de-452d-8eac-9cc9dd9947a0 - 39 - in: getters to not be called 3 
8. access getters more than it should https://playground.solidjs.com/anonymous/fe232aba-ecd7-47ff-83f2-2fbc0633b858 - 40 - in: getters to not be called 3.1
9. access getters more than it should 2 https://playground.solidjs.com/anonymous/cd367536-5769-4083-9cba-90d2ba3c9aa3 - 41 - in: getters to not be called 4 

10. problems with getters defined in classes 

- https://playground.solidjs.com/anonymous/46e40618-0655-426b-81b0-4d313dad470d - 49 - read and set class
-  https://playground.solidjs.com/anonymous/10f2b817-5eff-4ec9-b3c9-e6a5fe15dc26 - 51 - read and set inside class
- https://playground.solidjs.com/anonymous/deb37164-7105-438e-8657-99ab78f11a37 - 52 - read and set inside extended class
- https://playground.solidjs.com/anonymous/488709da-0ea6-4574-aaab-3e1ccf707385 - 53 - read and set inside extended x2 class 

11. doesnt react to hasOwnProperty https://playground.solidjs.com/anonymous/da5b1fb6-2914-4fdb-9bb0-ea65f8bdec38 - 55 - reacts to hasOwnProperty 
12. throws when redefining hasOwnProperty https://playground.solidjs.com/anonymous/78fb154e-388f-4f45-9335-07e855e9d644 - 79 - returns unproxied   properties 

13. deleting test 
- https://playground.solidjs.com/anonymous/8b1eaf25-9e8c-4d1b-b17b-03e542698d7c - 85 - supports not reacting when deleting a shallow property that was undefined
- https://playground.solidjs.com/anonymous/3505ca38-e13e-466e-8005-612353810737 - 88 - supports not reacting when deleting a deep property that was undefined
14. seems like its over reacting https://playground.solidjs.com/anonymous/ea8df396-5b69-402d-89a1-67a149bc53d0 - 94 - supports reacting to own keys deep
15. deleting 
- https://playground.solidjs.com/anonymous/b25dba78-20d8-42f6-a725-37acf763b5de 108 - supports reacting to property checks when value is undefined, deleting 
- https://playground.solidjs.com/anonymous/9e72b099-a67d-4a26-9f2f-9b0938228b4f - 109 - supports reacting to property checks when value is undefined, deleting deep 

16. reacting to adding properties https://playground.solidjs.com/anonymous/d4304507-190a-4097-846d-a4429108e7d9 - 110 - supports reacting to property checks, adding 
17. reacting to adding properties deep https://playground.solidjs.com/anonymous/7c7af674-159c-4707-82be-5a831d30d21d - 111 - supports reacting to property checks, adding deep 
18. throws with frozen objects https://playground.solidjs.com/anonymous/3e6b2e6e-7b78-4730-be4b-7e33ca6160fa - 122 - should not observe non-extensible objects
19. reacting to properties from the prototype https://playground.solidjs.com/anonymous/da7072f0-6a88-4a03-bbb8-b5ab9a642562 - 131 - should observe properties on the prototype chain
20. reacts when value changes from NaN to NaN (they probably should be using equals, but well) https://playground.solidjs.com/anonymous/8986df2c-9631-4f49-880c-c2e819c81aa9 - 147 - should not be triggered when the value and the old value both are NaN
21. splice and object identity https://playground.solidjs.com/anonymous/4a0a23b2-2131-47ef-bb0c-d604abeea1f8 - 175 - array: sliced test
22. infinite loop https://playground.solidjs.com/anonymous/f5403e80-7f3f-438f-8b94-fb3b5c29c81a 187 - array: pushing in two separated effects doesnt loop
23. identity test https://playground.solidjs.com/anonymous/cd68b1b1-6a7d-425e-ad11-e3b901fd77c9 205 - array: observed value should proxy mutations to original 
24. array methods for searching wont find stuff  https://playground.solidjs.com/anonymous/9ded9dba-aa62-43f4-8277-82edd361f523 - 206 - array: identity methods should work
25. more array identity tests https://playground.solidjs.com/anonymous/8f421b42-da3e-4bd1-937a-06d261690a44 - 208 - array: internal array functions should search for the mutable versions of it

27. infinite loop in array https://playground.solidjs.com/anonymous/499412e1-b975-4240-82fc-c440a20325e0 - 224 - array: should avoid infinite recursive loops when use Array.push/unshift/pop/shift

28. array returns past value? something is wrong with batching? https://playground.solidjs.com/anonymous/4e566456-5001-4147-b73c-5e188ac64270 -  226 - array: vue array instrumentation: concat

29. Possible related to batching https://playground.solidjs.com/anonymous/b5d60b40-a558-4828-a586-168b12aed411 -  93 - supports reacting to own keys

### Your Example Website or App

https://pota.quack.uy/Reactivity/mutable-tests

### Steps to Reproduce the Bug or Issue

-

### Expected behavior

-

### Platform

- OS: [e.g. macOS, Windows, Linux]
- Browser: [e.g. Chrome, Safari, Firefox]
- Version: [e.g. 91.1]

### Additional context

https://github.com/potahtml/pota

A playground, testing all our libs at the same time is in 
https://pota.quack.uy/Reactivity/mutable-tests

**Comment by ryansolid** [comment]
This is a great list that are worth reviewing. Thank you.

**Comment by westbrma** [comment]
I just tested a Map in a class that is wrapped with createMutable, changes to the map are not reactive, this may also be the same for a Set object.

**Comment by ryansolid** [comment]
> I just tested a Map in a class that is wrapped with createMutable, changes to the map are not reactive, this may also be the same for a Set object. I was doing a direct comparison with Mobx, I am curious why createMutable is not promoted more, I never cared for all the boilerplate around using and setting values, mobx and createMutable abstract all that away from me and generally just works.

Yeah Map and Set aren't handled. Any special type of object would need pretty special handling because the change mechanisms are internalized. The APIs for change are specific not assignment. We'd need to know how to handle them specifically. Like in MobX they have special classes for those. But if those why not something else. `createMutable` in general can be dangerous because the ability to mutate is implicit at any depth. Like you can pass parts of it around and implicitly give the ability to mutate it. Big proponent of read/write segregation around here.

**Comment by westbrma** [comment]
Because of these limitation I have continued to use Mobx with Solid, however I came across a big bug today. You created an example here on using mobx with Solid [https://codesandbox.io/s/mobx-external-source-0vf2l?file=/index.js:943-957](url)
If you reference any reactive property inside the component constructor (the code before the JSX) it will run the entire component again including the constructor code. In the example just add this line console.log(timer.secondsPassed) in the constructor. 

Also regarding the danger of mutating deep state anywhere, I'm not sure I would call it implicit, using the = assignment operator is an explicit statement. No one bothers with all the setter boilerplate in backend code, why is it different in frontend code? Maybe because it will re-render UI components?

**Comment by fabiospampinato** [comment]
@westbrma I can't reproduce the problem in the playground: https://playground.solidjs.com/anonymous/fff085cd-0234-4958-bf82-7cad295847dc

While I can reproduce it in the codesandbox with the same code, so I guess Babel is configured improperly in that codesandbox or something.

**Comment by westbrma** [comment]
> @westbrma I can't reproduce the problem in the playground: https://playground.solidjs.com/anonymous/fff085cd-0234-4958-bf82-7cad295847dc
> 
> While I can reproduce it in the codesandbox with the same code, so I guess Babel is configured improperly in that codesandbox or something.

strange, however you can reproduce if you click export to zip in your playground, then do
`npm install`
`npm run start`

**Comment by trusktr** [comment]
@westbrma to make a reactive Map you'd have to make a subclass of `Map`, and extend all the methods to read and write to/from Solid signals (or a store). Here's an example:

https://github.com/lume/element-behaviors/blob/main/src/BehaviorMap.ts

**Comment by nihil-pro** [comment]
Some tests contradict each other, for example:

> https://playground.solidjs.com/anonymous/8b1eaf25-9e8c-4d1b-b17b-03e542698d7c - 85 - supports not reacting when deleting a shallow property that was undefined

> https://playground.solidjs.com/anonymous/b25dba78-20d8-42f6-a725-37acf763b5de 108 - supports reacting to property checks when value is undefined, deleting

**Comment by titoBouzout** [comment]
85 and 108 aren't contradicting. 

> https://playground.solidjs.com/anonymous/8b1eaf25-9e8c-4d1b-b17b-03e542698d7c - 85 - supports not reacting when deleting a shallow property that was undefined

This is about that `o.value` is `undefined` and by deleting it, it will still be `undefined`, so the effect can be skipped. 

> https://playground.solidjs.com/anonymous/b25dba78-20d8-42f6-a725-37acf763b5de 108 - supports reacting to property checks when value is undefined, deleting

This is about that the effect has the check `if('value' in o){ }`, in first run that condition will be `true`, but if you delete the property then that condition won't be `true` any more and the other branch has to run.

**Comment by nihil-pro** [comment]
> 85 and 108 aren't contradicting.
> 
> > https://playground.solidjs.com/anonymous/8b1eaf25-9e8c-4d1b-b17b-03e542698d7c - 85 - supports not reacting when deleting a shallow property that was undefined
> 
> This is about that `o.value` is `undefined` and by deleting it, it will still be `undefined`, so the effect can be skipped.
> 
> > https://playground.solidjs.com/anonymous/b25dba78-20d8-42f6-a725-37acf763b5de 108 - supports reacting to property checks when value is undefined, deleting
> 
> This is about that the effect has the check `if('value' in o){ }`, in first run that condition will be `true`, but if you delete the property then that condition won't be `true` any more and the other branch has to run.

Yes, you're right. 
Of course, we could track reading property value (`Proxy.get`) separately, and access to the property without reading value (`Proxy.has`, `Proxy.deleteProperty`, `Proxy.getOwnPropertyDescriptor`) separately, but it seems that this will only add complexity, and will not bring significant benefits.

**Comment by titoBouzout** [comment]
Yeah, I agree there's some complexity, in theory it's a matter of keeping track of keys and values separately. 

I suppose it depends on what do you expect from a primitive like this.  Ive aimed to be granular.

Keep in mind the list of tests on this issue include only the ones that are failing. The complete list 250+ including the ones that solid pass is here https://github.com/potahtml/docs/blob/master/src/pages/tests/mutable.test.js

**Comment by ryansolid** [comment]
Thanks for putting this together. This was useful as an audit checklist.

For Solid 2.0, `createMutable` is no longer part of the core API, so we are not treating these as compatibility requirements for that primitive. The mutable write surface for stores is now `setStore(draft => ...)`, with readonly store reads outside the setter and batched updates.

I went through the categories here against the new store implementation. Most of these either do not apply to the Solid 2.0 store contract, are intentional proxy identity tradeoffs, or are cases we do not plan to support as full JS object membrane behavior.

One issue did carry over as a real store correctness bug: inherited prototype getters on class/prototyped objects could read through the raw instance instead of the store proxy, so reads inside the getter were not tracked. That has been fixed by making inherited accessor getters on non-plain store objects execute through the store proxy, without adding prototype lookup to plain object/array stores.

A few explicit decisions from the audit:
- Reactive property presence should use `"key" in store`; `hasOwnProperty` is not a reactive tracking API.
- Store arrays will not bridge raw/proxy identity for methods like `includes` / `indexOf`; compare store values or explicit keys.
- Store setters/accessor setters are not a supported store contract.
- Store equality follows signal equality, so `NaN` is not special-cased locally.
- Setting an absent key to `undefined` remains an accepted edge case.

Given `createMutable` is not continuing in Solid 2.0 core and the relevant carry-over bug has been addressed, I’m going to close this.

**Comment by ryansolid** [comment]
Here is the full audit for anyone interested:

# Solid 2.0 Store Mutable Write Audit

Source issue: <https://github.com/solidjs/solid/issues/2112>

`createMutable` is not part of Solid 2.0, so the issue is not a direct compatibility checklist. The relevant question is whether the Solid 2.0 mutable write surface for stores (`setStore(draft => ...)`) has analogous proxy/write semantics problems.

The executable audit lives in:

- `packages/solid-signals/tests/store/createStoreMutableAudit.test.ts`

Current suspected failures are encoded with `test.fails`, not implementation changes. Passing cases are normal tests.

## Solid 2.0 Semantics Used For This Audit

- Store reads outside a setter are readonly.
- Store writes happen through the setter draft.
- Writes batch like signals; outside the setter, values and presence generally settle after `flush()`.
- Audit expectations are written against post-`flush()` behavior unless a test explicitly checks draft-time behavior.

## Findings

### 1. Setting `undefined`

Status: intentional / accepted edge case.

Issue category:

- `setting to undefined shouldn't delete the property`

Audit case:

- `setting undefined on an absent key preserves property presence`

Current behavior:

- `setStore(s => { s.value = undefined; })` on an absent key leaves `"value" in store` false after `flush()`.

Expected behavior under normal JS object mutation:

- Assigning `undefined` creates the own property; deletion requires `delete`.

Notes:

- Existing tests cover setting an already-present key to `undefined`.
- The uncovered edge is absent key plus `undefined` value.
- This is probably acceptable: Solid 2.0 no longer treats `undefined` as delete, but an absent key set to `undefined` is skipped by equality/write elision. Guarding this edge would add complexity and cost for a narrow case.

### 2. Deleting Missing Keys

Status: covered / passes.

Issue category:

- `deleting key that doesn't exists triggers reactivity`

Audit case:

- `deleting a missing key does not notify key iteration`

Current behavior:

- Deleting an absent key does not rerun an `Object.keys(store)` observer.

Assessment:

- Solid 2.0 store behavior is good for this adapted case.

### 3. Deleting Present `undefined` Keys

Status: covered / passes.

Issue category:

- `delete key with undefined value does trigger reactivity - in`

Audit case:

- `deleting an undefined key notifies presence checks`

Current behavior:

- A present key with value `undefined` is removed by `delete`.
- `"value" in store` updates from true to false after `flush()`.

Assessment:

- Solid 2.0 store behavior is good for this adapted case.

### 4. `hasOwnProperty`

Status: follow-up.

Issue category:

- `doesnt react to hasOwnProperty`

Audit case:

- `hasOwnProperty tracks property additions`

Current behavior:

- `store.hasOwnProperty("value")` reports correctly after mutation.
- An effect depending on `store.hasOwnProperty("value")` does not rerun when the property is added.

Question:

- Should `hasOwnProperty` be a reactive presence read like `in`, or should users rely on `"key" in store` / key iteration for reactive presence?
- Need to confirm what actually fires for `store.hasOwnProperty("value")`: the proxy `get` trap for `hasOwnProperty`, then the built-in method likely performs an own-property descriptor lookup rather than the `has` trap. This needs a focused proxy trap trace before deciding.

### 5. Accessor Setter Writes

Status: design question / probably not recommended.

Issue category:

- `writing to an object with setter/getter overwrites getter`
- class getter/setter failures

Audit case:

- `assigning through an accessor property calls its setter`

Current behavior:

- Writing `s.value = 1` to an own accessor property changes the reported `store.value` after `flush()`, but does not call the original setter to update backing state.

Expected behavior under normal JS object mutation:

- Assignment to an accessor with a setter should call the setter with `this` bound to the receiver.

Risk:

- Store override semantics can diverge from class/accessor invariants by shadowing accessor output instead of using the setter.

Notes:

- Stores generally should not rely on setters; the public store value is intentionally immutable and writes are constrained to `setStore`.
- The audited case is specifically assignment to an accessor during `setStore`, e.g. `setStore(s => { s.value = 1; })`.
- Open question: if setters are unsupported, should we document that accessor setters are not honored, warn in dev, or attempt to call setters only when the descriptor is local and cheap to identify?

### 6. Getter-Only Writes

Status: design question.

Issue category:

- `attempting to set a value when its only a getter access getter`
- `access getters more than it should`

Audit case:

- `setting a getter-only property does not invoke the getter repeatedly`

Current behavior:

- Writing to a getter-only property can cause an extra getter read during write handling.

Question:

- Should a draft assignment to a getter-only property create an override, fail silently, or mirror strict JS assignment semantics?
- Current behavior appears to create/return an override while also causing extra getter access during write handling.
- Preferred direction may be to mirror strict JS assignment semantics, but this needs a concrete decision because current store override behavior is different.

### 7. Objects Returned By Store Methods

Status: covered / passes for store graph values.

Issue category:

- `object returned by function call in mutable is not observed`

Audit case:

- `objects returned by store methods remain reactive when they come from the store graph`

Current behavior:

- A store method returning `this.child` returns the proxied child.
- Effects tracking `store.getChild().value` update when the child is mutated.

Scope:

- This only covers returned objects that originate from the store graph.
- A method that constructs and returns a fresh object may be outside store semantics.

### 8. Prototype Getters

Status: fixed with narrow custom-prototype accessor path.

Issue category:

- `problems with getters defined in classes`
- `reacting to properties from the prototype`

Audit case:

- `prototype getters track instance field updates`

Current behavior:

- Reading a prototype getter such as `store.doubled` now reflects `store.value` changes after `flush()` in the audited case.

Likely cause:

- Store `get` checks own descriptors on the selected store value path, but does not evaluate accessor descriptors through the prototype chain with the proxy receiver.

Notes:

- This was a real semantic bug for class/prototype getter reads.
- The fix is intentionally narrow: plain objects and arrays are classified as non-custom once at proxy creation, while custom-prototype targets only walk inherited descriptors when a property has no own descriptor.
- Projection stores seeded from another store are covered; the actual read uses `Reflect.get` on the selected store value so inner store proxies still participate.

### 9. Prototype Setters

Status: probably unsupported / design question.

Issue category:

- class setter/getter failures

Audit case:

- `prototype setters are honored by draft writes`

Current behavior:

- Draft assignment to a prototype setter does not call the setter in the audited case.

Expected behavior under normal JS object mutation:

- The prototype setter should update instance state via the receiver.

Notes:

- As with finding 5, supporting setters in stores may not be worth the complexity.
- If setters remain unsupported, this should probably be documented alongside accessor setter behavior rather than fixed piecemeal.

### 10. Array Identity Search With Raw Values

Status: intentional / won't fix.

Issue category:

- `array methods for searching wont find stuff`
- `internal array functions should search for the mutable versions of it`

Audit case:

- `array identity search methods match raw values from the source array`

Current behavior:

- `store.includes(rawItem)` is false when `rawItem` is the original object used to create the store.

Question:

- Should store arrays bridge raw/proxy identity for `includes`, `indexOf`, and `lastIndexOf`, or should identity comparisons consistently require store-proxy values?

Decision:

- Won't fix. Store arrays are not expected to re-implement array methods to bridge raw/proxy identity. Callers should compare against values read from the store or use explicit keys.

### 11. Array Identity Search With Store Values

Status: covered / passes.

Issue category:

- array identity methods

Audit case:

- `array identity search methods match store values`

Current behavior:

- `const item = store[0]; store.includes(item)` works.

Assessment:

- Proxy identity is internally consistent when callers use values read from the store.

### 12. Array Mutator Recursion

Status: covered / passes for audited case.

Issue category:

- `array: pushing in two separated effects doesnt loop`
- `array: should avoid infinite recursive loops when use Array.push/unshift/pop/shift`

Audit case:

- `array mutators do not recursively notify themselves`

Current behavior:

- An effect that pushes once when length is zero settles at length one and does not loop.

Scope:

- This is a minimal recursion check, not a full port of every array mutator case from the issue.

### 13. `NaN` Equality

Status: intentional / won't fix unless signal equality changes.

Issue category:

- `reacts when value changes from NaN to NaN`

Audit case:

- `NaN to NaN does not notify observers`

Current behavior:

- Setting `NaN` over `NaN` reruns observers.

Likely cause:

- Store property signals use the same equality behavior as signals, which compares with `===`.

Question:

- Should stores use SameValueZero / `Object.is`-style equality locally, or is `NaN !== NaN` consistent with signal semantics and therefore intentional?

Decision:

- Won't fix locally. If signals use `===` semantics, stores should not diverge just for `NaN`.

## Not Yet Ported From The Issue

These need separate focused tests before classification:

- Frozen non-extensible root values and nested values beyond getter returns.
- Redefining `hasOwnProperty` as a user property.
- Full class inheritance chain cases.
- `splice`, `slice`, and object identity through array replacement.
- `concat` and batching around array method return values.
- Full `ownKeys` overreaction scenarios from the external suite.

## Suggested Review Order

1. Decide the intended store semantics for absent-key `undefined` writes.
2. Decide whether `hasOwnProperty` is part of the reactive API surface.
3. Decide whether accessor/prototype setters should follow normal JS assignment semantics inside setter drafts.
4. Decide whether store arrays should bridge raw/proxy identity in search methods.
5. Decide whether store property equality should diverge from signal equality for `NaN`.
6. Extend the audit to the not-yet-ported array and inheritance cases before implementing fixes.

## Expensive Case Investigation

### Finding 4: `hasOwnProperty`

Investigation:

- `store.hasOwnProperty("value")` triggers the proxy `get` trap for `hasOwnProperty`.
- The built-in method then checks ownership through `getOwnPropertyDescriptor`, not the `has` trap.
- `Object.prototype.hasOwnProperty.call(store, "value")` also uses `getOwnPropertyDescriptor`.
- The current reactive presence machinery is attached to the `has` trap (`"value" in store`), so `hasOwnProperty` does not naturally share that tracking.

Cost / implementation shape:

- Making this reactive through `getOwnPropertyDescriptor` would risk making all descriptor reads reactive, which is broader than needed.
- The narrower option is a `get` special-case for `hasOwnProperty` that returns a store-aware function backed by the same presence nodes as `in`, or separate own-presence nodes if inherited presence must differ.

Recommendation:

- Do not make descriptor reads broadly reactive.
- Document that reactive presence uses `in`, not `hasOwnProperty`.
- Do not support reactive `hasOwnProperty` unless there is a later strong user-facing reason.

### Finding 6: Getter-Only Writes

Investigation:

- Current write prep reads `state[property]` before deciding how to store the override.
- For getter-only properties, that means the getter can run during assignment.
- The store then stores an override rather than following strict JS assignment behavior.

Cost / implementation shape:

- Fixing this does not require read-path cost if handled only in the `set` trap.
- Avoiding the extra getter call requires checking descriptors before `prepareStoreWrite`, because `prepareStoreWrite` currently reads `state[property]`.
- An unconditional own descriptor lookup would affect every draft write, not every read.
- A lower-cost option is to cache whether a store target has own accessor descriptors, then only pay the descriptor lookup for objects that might need it.
- That would let the setter decide:
  - getter-only descriptor: no-op / dev warn / strict-like failure policy
  - accessor with setter: unsupported warning rather than implicit override

Recommendation:

- Treat this as a write-path design decision, not a read-path performance issue.
- Prefer no broad support for setters, but consider a dev warning for accessor writes because current override behavior is surprising.
- If we mirror strict JS assignment semantics, the most honest behavior for getter-only writes in module code would be to fail the assignment; however, that may be too harsh for store ergonomics. A dev warning plus no-op may be the pragmatic compromise.

### Finding 8: Prototype Getters

Investigation:

- Current `get` only checks own descriptors with `Object.getOwnPropertyDescriptor(storeValue, property)`.
- If no own descriptor exists, `storeValue[property]` evaluates prototype getters with `this` bound to the raw target.
- That means a prototype getter like `get doubled() { return this.value * 2; }` reads raw `value`, bypassing store tracking.
- `Reflect.get(storeValue, property, receiver)` would call the getter with `this` bound to the proxy, which would track `this.value`.
- This is not only "class objects" in the language sense. It applies to any store object whose useful accessor is inherited from its prototype. Classes are the common way this happens.

Cost / implementation shape:

- Replacing normal reads with `Reflect.get` everywhere would add cost to the hot path.
- Walking the prototype chain on every missing own descriptor is much more expensive than direct own reads.
- A possible narrow implementation is a gated prototype-accessor lookup:
  - only after own descriptor miss
  - skip arrays and `Object.prototype`
  - use `Reflect.get(storeValue, property, receiver)` only for inherited accessor properties
- A cheaper semantic compromise is to only consider prototype accessors when the target's prototype is neither `Object.prototype` nor `Array.prototype`, because plain object stores dominate the hot path and class/prototyped objects are the edge case.

Recommendation:

- This was the only getter/setter-related item that justified a source fix.
- Keep the scope limited to inherited getter reads; do not expand this into setter support or a full object membrane.

### Finding 5 / 9: Setters

Investigation:

- Own and prototype setters are both currently bypassed by store override writes.
- Supporting them fully would move stores closer to a JS object membrane, which conflicts with the optimized plain-data store model.

Recommendation:

- Do not support setters as a store contract.
- Consider a dev-only warning for writes to accessor descriptors if it can be detected on the write path without affecting reads.
